### PR TITLE
Report error in `killTasks`.

### PR DESCRIPTION
**Pull request by jeschkies**
This relates to #4191. However, I couldn't test the reported case.

There are a few comments on code I'm unsure about.

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/api/TaskKiller.scala [hunk 1]
```diff
@@ -34,32 +36,40 @@ class TaskKiller @Inject() (
...
+
+        val processNextTask = stateOpProcessor.process(TaskStateOp.ForceExpunge(nextTask.taskId))
+          .map(_ => Done)
+          .recover {
+            case NonFatal(cause) =>
+              log.info("Failed to expunge {}, got: {}", Array[Object](nextTask.taskId, cause): _*)
+              Done // Note, this error is swallowed.
+          }
+
+        await(processNextTask)
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/TaskKiller.scala]: This is supposed to make it more explicit that we purge sequentially and not in parallel.

File: src/test/scala/mesosphere/marathon/test/CaptureLogEvents.scala [hunk 2]
```diff
@@ -21,6 +21,12 @@ object CaptureLogEvents {
     private[this] def rootLogger: Logger = LoggerFactory.getLogger(org.slf4j.Logger.ROOT_LOGGER_NAME).asInstanceOf[Logger]
 
     def appendToRootLogger(): Unit = {
+      // Waiting for Logback to be ready.
+      // See https://gist.github.com/mbknor/34944ea4589a5fc6974c
+      // Should actually have been fixed with http://jira.qos.ch/browse/SLF4J-167
+      while (LoggerFactory.getILoggerFactory.isInstanceOf[org.slf4j.helpers.SubstituteLoggerFactory]) {
+        Thread.sleep(50)
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/test/CaptureLogEvents.scala]: Not happy about this busy waiting.

File: src/main/scala/mesosphere/marathon/api/TaskKiller.scala [hunk 3]
```diff
@@ -34,32 +36,40 @@ class TaskKiller @Inject() (
       case Some(app) =>
         checkAuthorization(UpdateRunSpec, app)
 
+        // TODO: We probably want to pass the execution context as an implcit.
         import scala.concurrent.ExecutionContext.Implicits.global
-        taskTracker.appTasks(appId).flatMap { allTasks =>
+
+        async {
```
Comment (unterstein) [inline src/main/scala/mesosphere/marathon/api/TaskKiller.scala]: Why using async await and not just asking the actor?

```
     taskTracker.appTasks(appId).flatMap { allTasks =>
          val foundTasks = findToKill(allTasks)

          if (wipe) expunge(foundTasks)

          val launchedTasks = foundTasks.filter(_.launched.isDefined)
          if (launchedTasks.nonEmpty) service.killTasks(appId, launchedTasks).map(killed => killed.tasks)
          else Future.successful(Seq.empty)
        }
```

File: src/main/scala/mesosphere/marathon/api/TaskKiller.scala [hunk 4]
```diff
@@ -34,32 +36,40 @@ class TaskKiller @Inject() (
...
-              foundTasks
-            } else foundTasks
-          }
+
+          if (wipe) expunge(foundTasks)
+
+          val launchedTasks = foundTasks.filter(_.launched.isDefined)
+          if (launchedTasks.nonEmpty) await(service.killTasks(appId, launchedTasks))
+          // TODO: For some reasons tasks returned by service.killTasks !=
+          // foundTasks in some cases. What should be returned?
```
Comment (unterstein) [inline src/main/scala/mesosphere/marathon/api/TaskKiller.scala]: That`s correct IMO. Only killed tasks should be returned.

File: src/main/scala/mesosphere/marathon/api/TaskKiller.scala [hunk 5]
```diff
@@ -34,32 +36,40 @@ class TaskKiller @Inject() (
...
       case None => Future.failed(UnknownAppException(appId))
     }
   }
 
-  private[this] def expunge(tasks: Iterable[Task])(implicit ec: ExecutionContext): Future[Unit] = {
-    tasks.foldLeft(Future.successful(())) { (resultSoFar, nextTask) =>
-      resultSoFar.flatMap { _ =>
+  private[this] def expunge(tasks: Iterable[Task])(implicit ec: ExecutionContext): Unit = {
+    tasks.foreach { nextTask =>
+      async {
```
Comment (unterstein) [inline src/main/scala/mesosphere/marathon/api/TaskKiller.scala]: Why using async await?

File: src/test/scala/mesosphere/marathon/api/TaskKillerTest.scala [hunk 6]
```diff
@@ -137,7 +164,8 @@ class TaskKillerTest extends MarathonSpec
     when(f.tracker.appTasks(appId)).thenReturn(Future.successful(tasksToKill))
     when(f.stateOpProcessor.process(stateOp1)).thenReturn(Future.successful(TaskStateChange.Expunge(runningTask)))
     when(f.stateOpProcessor.process(stateOp2)).thenReturn(Future.successful(TaskStateChange.Expunge(reservedTask)))
-    when(f.service.killTasks(appId, launchedTasks)).thenReturn(launchedTasks)
+    when(f.service.killTasks(appId, launchedTasks))
+      .thenReturn(Future.successful(MarathonSchedulerActor.TasksKilled(appId, launchedTasks)))
```
Comment (unterstein) [inline src/test/scala/mesosphere/marathon/api/TaskKillerTest.scala]: Should return only launched tasks (only launched tasks will be killed)

File: src/test/scala/mesosphere/marathon/test/CaptureLogEvents.scala (hunk 2 above)
Comment (unterstein) [inline src/test/scala/mesosphere/marathon/test/CaptureLogEvents.scala]: ouh, I understand why you are unhappy :-). Whats happening there?

File: src/main/scala/mesosphere/marathon/api/TaskKiller.scala (hunk 5 above)
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/TaskKiller.scala]: I want to make it explicit that the tasks are not processed in parallel but rather sequentially.

Comment (unterstein) [inline src/main/scala/mesosphere/marathon/api/TaskKiller.scala]: Was this a requirement?

File: src/main/scala/mesosphere/marathon/api/TaskKiller.scala (hunk 4 above)
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/TaskKiller.scala]: Then I have to fix it.

File: src/main/scala/mesosphere/marathon/api/TaskKiller.scala (hunk 3 above)
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/TaskKiller.scala]: I talked to Jason about Scala Async and agreed that it has advantages over nested flatMaps or Scala's for comprehensions. E.g. it easier to reason about and avoids eird things like Future.sucessful(...) or the map.

File: src/main/scala/mesosphere/marathon/api/TaskKiller.scala (hunk 4 above)
Comment (unterstein) [inline src/main/scala/mesosphere/marathon/api/TaskKiller.scala]: Would suggest something like
```
if (launchedTasks.nonEmpty) await(service.killTasks(appId, launchedTasks)).map(killed => killed.tasks)
else Seq.empty
```

Maybe @meichstedt will add his opinion :)

File: src/main/scala/mesosphere/marathon/api/TaskKiller.scala (hunk 5 above)
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/TaskKiller.scala]: The original code is and Jason said that there have been some limitations.

File: src/main/scala/mesosphere/marathon/api/TaskKiller.scala (hunk 4 above)
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/TaskKiller.scala]: This change lets `TaskKillerTest "kill with wipe will kill running and expunge all"` fail. The expected return value is all killed tasks *and* all expunged tasks.

File: src/main/scala/mesosphere/marathon/api/TaskKiller.scala (hunk 5 above)
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/TaskKiller.scala]: I reverted the change of the expunge method. My change was not completely sequential.

**Comment by jeschkies** [comment]
Added a small comment. @aquamatthias @meichstedt Could please verify.

**Comment by meichstedt** [comment]
@jeschkies @aquamatthias @unterstein tl;dr: LGTM.

Long version: Killing a resident task will kill the running Mesos task, but not expunge its metadata (because we might eventually want to launch a new task using the same reservations/persistent volumes). That's what the `wipe` flag is there for: if `wipe==true`, the metadata will be expunged _and_ the task killed if it's launched (a workaround because there's no StorageManagement providing this functionality). It is definitely possible that a task is found, expunged but currently not launched, so there won't be a kill request to Mesos. Yet, since the user passed its ID and asked to wipe the metadata, she should receive the taskId as a kind of ACK.

**Comment by meichstedt** [comment]
@jeschkies please rebase

**Comment by jeschkies** [comment]
@meichstedt I guess we can merge now : )

**Comment by meichstedt** [comment]
👍

**Pull request merged**